File: Scala/0019.scala
/**
 * Definition for singly-linked list.
 * class ListNode(_x: Int = 0, _next: ListNode = null) {
 *   var next: ListNode = _next
 *   var x: Int = _x
 * }
 */
object Solution {
    def removeNthFromEnd(head: ListNode, n: Int): ListNode = {
        // Create a dummy node which helps in edge cases like removing the head
        val dummy = new ListNode(0)
        dummy.next = head
        
        // Initialize the two pointers
        var first: ListNode = dummy
        var second: ListNode = dummy
        
        // Move `first` pointer n+1 steps ahead
        for (_ <- 0 to n) {
            first = first.next
        }
        
        // Move both pointers until `first` reaches the end
        while (first != null) {
            first = first.next
            second = second.next
        }
        
        // `second.next` is the node to be removed
        second.next = second.next.next
        
        // Return the new head which is the next of dummy
        dummy.next
    }
}

File: C/0019.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

// Function to remove the nth node from the end of the list
struct ListNode* removeNthFromEnd(struct ListNode* head, int n) {
    // Create a dummy node to simplify edge cases (like removing the head node)
    struct ListNode dummy;
    dummy.next = head;
    
    // Initialize two pointers
    struct ListNode* fast = &dummy;
    struct ListNode* slow = &dummy;
    
    // Move fast pointer n+1 steps ahead so that the gap between fast and slow is n
    for (int i = 0; i <= n; i++) {
        fast = fast->next;
    }
    
    // Move both fast and slow pointers until fast reaches the end
    while (fast != NULL) {
        fast = fast->next;
        slow = slow->next;
    }
    
    // Remove the nth node from end
    struct ListNode* nodeToRemove = slow->next;
    slow->next = slow->next->next;
    
    // Free the memory of the removed node
    free(nodeToRemove);
    
    return dummy.next;
}

// Helper function to create a new node
struct ListNode* createNode(int val) {
    struct ListNode* newNode = (struct ListNode*)malloc(sizeof(struct ListNode));
    newNode->val = val;
    newNode->next = NULL;
    return newNode;
}

// Helper function to print the list
void printList(struct ListNode* head) {
    while (head != NULL) {
        printf("%d ", head->val);
        head = head->next;
    }
    printf("\n");
}